Google.Api.Gax.Grpc.Tests/ApiCallTest.cs

```diff
@@ -132,8 +136,310 @@ public void WithGoogleRequestParam()
+            CallSettingsTest.AssertRoutingHeader(asyncCallSettings, expectedHeader);
+        }
+
+        /// <summary>
+        /// Extracting a single routing header key-value pair by matching
+        /// several conflictingly named path templates on (parts of) a single request
+        /// field. The last template to match "wins" the conflict.
+        /// </summary>
+        [Theory]
+        [InlineData("projects/100", "table_name=projects/100")]
+        [InlineData("projects/100/shards/300", "table_name=projects/100")]
+        [InlineData("projects/100/instances/200", "table_name=projects/100/instances/200")]
+        [InlineData("projects/100/instances/200/shards/300", "table_name=projects/100/instances/200")]
+        [InlineData("orgs/1/projects/100/instances/200", null)]
+        [InlineData("foo", null)]
+        public void WithExtractedGoogleRequestParam_OverlappingPatterns(string fieldValue, string expectedHeader)
+        {
+            const string headerName = "table_name";
+
+            CallSettings syncCallSettings = null;
+            CallSettings asyncCallSettings = null;
+            var call0 = new ApiCall<ExtractedRequestParamRequest, SimpleResponse>(
+                (req, cs) => { asyncCallSettings = cs; return null; },
+                (req, cs) => { syncCallSettings = cs; return null; },
+                null);
+
+            // call corresponding to the following routing parameters:
+            // { field: "table_name", path_template: "{routing_id=projects/*}/**" }
+            // { field: "table_name", path_template: "{routing_id=projects/*/instances/*}/**" }
+            var call1 = call0
+                .WithExtractedGoogleRequestParam(
+                    new HeaderParameterExtractor<ExtractedRequestParamRequest>(headerName,
+                            new Regex("^(?<routing_id>projects/[^/]+)(?:/.*)?$"), request => request.TableName)
+                        .Add(headerName,
+                            new Regex("^(?<routing_id>projects/[^/]+/instances/[^/]+)(?:/.*)?$"), request => request.TableName));
+            call1.Sync(new ExtractedRequestParamRequest { TableName = fieldValue}, null);
+            call1.Async(new ExtractedRequestParamRequest { TableName = fieldValue }, null);
+
+            CallSettingsTest.AssertRoutingHeader(syncCallSettings, expectedHeader);
+            CallSettingsTest.AssertRoutingHeader(asyncCallSettings, expectedHeader);
+        }
+
+        /// <summary>
+        /// Extracting multiple routing header key-value pairs by matching
+        /// several non-conflicting path templates on (parts of) a single request field.
+        /// Make the templates strict, so that if the `table_name` does not
+        /// have an instance information, nothing is sent.
+        /// </summary>
+        [Theory]
+        [InlineData("projects/100/instances/200/tables/300", "project_id=projects/100&instance_id=instances/200")]
```

Comment:
This test is sensitive to the order in which they're reported by the dictionary - this sort of thing is why I think would be better to just keep a list of parameter/value pairs, or put everything into the parameter extractor (it could just produce the complete header value).

Reply:
I've fixed the test to not be order sensitive. Keeping the string here because (a) it is simpler to read and (b) parity with Ruby tests